rs/execution_environment/src/scheduler/round_schedule.rs

```diff
@@ -229,4 +231,77 @@ impl RoundSchedule {
+        is_first_iteration: bool,
+        rank: usize,
+    ) {
+        let full_message_execution = match canister.next_execution() {
+            NextExecution::None => true,
+            NextExecution::StartNew => false,
+            // We just finished a full slice of executions.
+            NextExecution::ContinueLong | NextExecution::ContinueInstallCode => true,
+        };
+        let scheduled_first = is_first_iteration && rank == 0;
+
+        if full_message_execution || scheduled_first {
+            // The very first canister is considered to have a full execution round for
+            // scheduling purposes even if it did not complete within the round.
+            canister.scheduler_state.last_full_execution_round = round_id;
+
+            // For now the priority for full execution is always 100.
+            // In future we might use a proportion to the actual load.
+            executed_priority_ids.insert(canister.canister_id(), 100.into());
+        }
+    }
+
+    /// Finishes inner round execution.
+    pub(crate) fn finish_inner_round(
+        &self,
+        canister_states: &mut BTreeMap<CanisterId, CanisterState>,
+        executed_priority_ids: BTreeMap<CanisterId, AccumulatedPriority>,
+    ) {
+        let scheduler_cores = self.scheduler_cores;
+        let compute_capacity_percent =
+            SchedulerImpl::compute_capacity_percent(scheduler_cores) as i64;
+        let number_of_canisters = canister_states.len();
+        let multiplier = (scheduler_cores * number_of_canisters).max(1) as i64;
+
+        // Charge canisters for the priority executed in the previous round.
+        let mut total_executed_priority = 0;
+        for (canister_id, executed_priority) in executed_priority_ids {
+            if let Some(canister) = canister_states.get_mut(&canister_id) {
+                total_executed_priority += executed_priority.get() * multiplier;
+                // If priority credit is non-zero, use it instead of accumulated priority.
+                if canister.scheduler_state.priority_credit != 0.into() {
+                    canister.scheduler_state.priority_credit += executed_priority * multiplier;
+                } else {
+                    canister.scheduler_state.accumulated_priority -= executed_priority * multiplier;
+                }
+            }
+        }
+
+        // Scale the executed priority to compute capacity.
+        let executed_multiplier = total_executed_priority / compute_capacity_percent;
+        // Distribute the executed priority across canisters according to their allocations.
+        let mut total_allocated = 0;
+        for (_canister_id, canister) in canister_states.iter_mut() {
+            let allocated = canister.scheduler_state.compute_allocation.as_percent() as i64
+                * executed_multiplier;
+            total_allocated += allocated;
+            canister.scheduler_state.accumulated_priority += allocated.into();
+        }
```

Comment:
I'm really not sure what is happening here. It seems to me like if we executed 20 canisters and some canister has a compute allocation of 70, then we credit it with something like 20 / 3 * 70 priority? If so, this is wrong. What if this goes on indefinitely? This canister will accumulate huge priority (~5x more than it can possibly spend).

Given the example above, we should simply pretend as if we had 20 cores / threads instead of 4 and apply the same logic as before: give the canister with the 70 compute allocation 70 priority; and split the remaining 20 * 100 - 70 priority equally across all canisters (including the 70 compute allocation one).

Reply:
Discussed offline, now we just use a normal multiplier: https://github.com/dfinity/ic/pull/1782/commits/89b8a6255eed61b1e400bed89a37c132ae3c7402